### PR TITLE
Custom Data: Add SalesAccount and CustomData [AG-1371]

### PR DESCRIPTION
**Pull request by moreira**
AG-1371

### Problem

We need to expose custom data for account groups through some of our external APIs.

### Solution

Account groups are exposed as business locations and sales accounts. We decided to expose custom data for the salesAccounts endpoints.

Custom data can be accessed via GET through:
- GET `/platform/salesAccounts/{agid}/customData` and `/platform/salesAccountCustomData/{agid}`, returns only custom data for account
- GET `/platform/salesAccounts/{agid}`, returns sales account with the custom fields in the `included` relationships payload
- GET `/platform/salesAccounts/{agid}/relationships/customData`, only returns relationship ID identified by account group ID

Updates via:
- PATCH `/platform/salesAccountCustomData/AG-33QK28ZNJP`
with a payload like:
```javascript
{
    "data": {
        "id": "AG-8RWGZ7JVMV",
        "type": "salesAccountCustomData",
        "attributes": {
            "partnerFields": [
                {
                    "fieldId":"FieldID-054034ab-0ee7-411b-b836-066cfc64b4d9",
                    "integerValue":50
                },
                {
                    "fieldId": "FieldID-9dd448ca-90d0-4b15-ba72-70d640bf4b99",
                    "stringValue": "test PATCH v2",
                },
                 {
                    "fieldId": "FieldID-195cd902-a141-40cb-8dd5-893a339743bb",
                    "dateValue": "2021-12-17T08:33:33Z",
                }
            ]
        }
    }
}
```

Data returned in a GET is of the format:
```javascript
{
    "data": {
        "type": "salesAccountCustomData",
        "id": "AG-8RWGZ7JVMV",
        "attributes": {
            "partnerFields": [
                {
                    "fieldId": "FieldID-195cd902-a141-40cb-8dd5-893a339743bb",
                    "title": "Foundation date",
                    "description": "Enter a foundation date.",
                    "dateValue": "2021-12-17T08:33:33Z",
                    "fieldType": "date"
                },
                {
                    "fieldId": "FieldID-7e686ac9-e28c-4f3d-b56a-9cb0d29e0be2",
                    "title": "Lead Tier",
                    "description": "Select a lead tier.",
                    "fieldType": "dropdown"
                },
                {
                    "fieldId": "FieldID-7fbe25b5-9cca-4c7a-9e04-34f93a2a4abd",
                    "title": "Revenue",
                    "description": "Enter revenue",
                    "fieldType": "integer"
                },
                {
                    "fieldId": "FieldID-9dd448ca-90d0-4b15-ba72-70d640bf4b99",
                    "title": "Conversion Point",
                    "description": "Enter a conversion point.",
                    "stringValue": "test PATCH v2",
                    "fieldType": "string"
                },
                {
                    "fieldId": "FieldID-054034ab-0ee7-411b-b836-066cfc64b4d9",
                    "title": "Number of employees.",
                    "description": "Enter your number of employees.",
                    "integerValue": 50,
                    "fieldType": "integer"
                }
            ]
        }
    }
}
```
It will contain all custom fields, even if they are empty (because they have never been set)

@vendasta/np-easy AG-1371

### CODE REVIEW

**Comment by richard-rance** [comment]
Please have a look at [this doc](https://vendasta.jira.com/wiki/spaces/API/pages/1651769533/Pre+Release+Checks#api-gateway-docs) for a list of things that should be checked.

File: openapi/platform/platform.yaml [hunk 1]
```diff
@@ -1423,6 +1423,46 @@ paths:
...
+                    type: salesAccounts
+                    attributes:
+                      tags:
+                        - tag1
+                        - tag2
+                      customerIdentifier: User-defined-id-123123
+                      name: Company Example
+                      address:
+                        line1: 109 8th Street E.
+                        line2: Suite 23
```
Comment (etarasoff-va) [inline openapi/platform/platform.yaml:1439]: If you're going to use a real address in docs, I'd use a famous landmark like the CN Tower. `290 Bremner Blvd, Toronto, ON M5V 3L9`

File: openapi/platform/platform.yaml [hunk 2]
```diff
@@ -1660,6 +1886,384 @@ paths:
...
+    options:
+      summary: ''
+      operationId: options-salesAccountCustomDataID-by-salesAccount
+      responses:
+        '204':
+          description: No Content
+      description: 'Used solely for [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) the OPTIONS request returns the list of possible HTTP methods and other headers that browsers use to protect user''s security. You should not call this operation directly. '
+      tags:
+        - Sales Account Custom Data
+  '/salesAccounts/{id}/customData':
```
Comment (richard-rance) [inline openapi/platform/platform.yaml:2014]: How should we describe the difference between `/salesAccounts/{id}/customData` and `/salesAccountCustomData/{id}`?
Should we allow updates to this path as well so they are the exact same?

**Review by richard-rance — COMMENTED** [review]
I left a few comments as I reviewed. When I got to some more complex issues with the OAS I switched to just making the changes so there are only a few for you to do. 

By the way you can preview what your docs will look like at this hidden URL https://developers.vendasta.com/platform/branches/np-ag-1371-reform-sales-account/ZG9jOjEwMTkzMDgz-change-log

File: openapi/platform/platform.yaml [hunk 3]
```diff
@@ -1446,6 +1525,7 @@ paths:
 
         The following members must be populated during creation:
         - `relationships.businessPartner.data.id`
+        - `relationships.salesPeople.data.[].id` - At least one user's id of a sales person must be provided
```
[the remark this answers is not in the document]
Comment (moreira) [reply openapi/platform/platform.yaml:1529]: I thought that was the difference between business Location and Sales Account. But if salesperson is optional, ok

File: openapi/platform/platform.yaml (hunk 2 above)
Comment (moreira) [reply openapi/platform/platform.yaml:2014]: I think we discussed the update on `/salesAccounts/{id}/customData`. We would have to add support for that in the api2go, the problem is that we 'd have to add support for a lot of types of relationship updates. Thats the reason we left salesAccountCustomData as a top-level entity.

File: openapi/platform/platform.yaml [hunk 4]
```diff
@@ -2857,7 +3534,144 @@ components:
...
+                description: 'Field identifier, can be retrived by listing the attributes'
+              stringValue:
+                type: string
+                description: Required when the field type is 'string'
+              integerValue:
+                type: integer
+                description: Required when the field type is 'integer'
+              dateValue:
+                type: string
+                description: Required when the field type is 'date'
```
[the remark this answers is not in the document]
Comment (moreira) [reply openapi/platform/platform.yaml]: Fixing here: https://github.com/vendasta/api-gateway/pull/126

File: openapi/platform/platform.yaml [hunk 5]
```diff
@@ -2798,7 +3453,7 @@ components:
                       id:
                         type: string
                         pattern: ^U-
-                        description: The user id of a salesperson
+                        description: The user id of a salesperson. Can be retrieved using the platform or users endpoint
```
[the remark this answers is not in the document]
Comment (moreira) [reply openapi/platform/platform.yaml:3445]: I thought about the partner-center (platform) and the `/platform/users` (I'm suggestion ways to discover the `id` so I can't suggest `GET platform/users/{id}`).

I'm gonna rephrase it and add the include option

File: openapi/platform/platform.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (richard-rance) [reply openapi/platform/platform.yaml:1529]: The difference is the salesPeople, tags and other partner employee only fields only exist on the sales account. Please take this line out.

File: openapi/platform/platform.yaml (hunk 5 above)
[the remark this answers is not in the document]
Comment (richard-rance) [reply openapi/platform/platform.yaml:3445]: Ah. The list users endpoint with some filters is the recommended way to find the user id.

File: openapi/platform/platform.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (moreira) [reply openapi/platform/platform.yaml:1529]: Fixed by https://github.com/vendasta/api-gateway/pull/128

File: openapi/platform/platform.yaml [hunk 6]
```diff
@@ -2071,9 +2070,7 @@ paths:
       description: |-
         [Status](https://developers.vendasta.com/platform/ZG9jOjEwMTU2NTYy-versioning): `Trusted Tester`
 
-        Returns custom metadata about the sales account.
-
-        Attributes can be set up at the partner administration page in Partner Center.
+        This endpoint acts like a alias to [Get Sales Account Custom Data by ID endpoint](/platform/branches/np-ag-1371-reform-sales-account/b3A6MzYxMTM5MTY-get-sales-account-custom-data-by-id). It returns custom fields about a sales account.
```
Comment (richard-rance) [inline openapi/platform/platform.yaml:2073]: The absolute link will break when we merge to master and delete this branch.

```suggestion
        This endpoint acts like a alias to [Get Sales Account Custom Data by ID endpoint](b3A6MzYxMTM5MTY-get-sales-account-custom-data-by-id). It returns custom fields about a sales account.
```

Comment (richard-rance) [reply openapi/platform/platform.yaml:2073]: I have updated this.

**Review by richard-rance — APPROVED** [review]
This looks good. After merging please update the copy of the yaml files in the api-gateway.

In the next PR I would like a guide showing partners how to use the new endpoints.

**Pull request merged**